### ISSUE TITLE
Performing HTTPS request in a custom filter

### ISSUE THREAD

**Issue by brado** [Ettercap/ettercap#1112]
The use case I am trying to accomplish is as follows: 

Description: Using ARP-poisoning, create a MITM situation where HTTP requests from the targeted client can be intercepted by Ettercap and within (presumably) a custom filter, it requests (over HTTPS) from the server, and then returns the response (slightly altered) to the client. So in general, the request flow: 

REQUEST FLOW: Client browser -- HTTP --> Ettercap (filter) -- HTTPS --> Server web site
RESPONSE FLOW: Server web site -- HTTPS --> Ettercap (filter) -- HTTP --> Client browser

I've been considering various options such as dropping the original request, executing a curl request in a shell, etc. but I'm not clear (even by looking at example filters and Ettercap source) what the proper approach would be to make this happen in a filter. For instance, a few of the questions: 

- If a curl request is executed in a shell, what happens to the response -- does it (or when does it) return to the curl command after interception by Ettercap again (which would seem to have deadlock possibilities, if a filtered request packet has to wait on a filtered response packet to return).
- if a packet is intercepted from the client in a filter, is there a way to immediately return a response back to the client (vs forward the packet on to its original destination)? 

Help is appreciated. Thx.

**Comment by koeppea** [comment]
I'm afraid this is not possible with Ettercap.
I can't even imagine a scenario where such a behavior would be useful.
Can you elaborate more on the use-case?

**Comment by brado** [comment]
Sure...take sslstrip -- what it attempts to do is when HSTS is not present, take a straight unencrypted HTTP response and rewrite any URLs within the body to use HTTP protocol rather than HTTPS. That made for a rather subtle but insecure HTTP communication through a MITM with credentials / data intercepted, prior to various warnings that now appear in current versions of browsers. In Ettercap, breaking SSL/TLS is visibly disruptive to the browser experience.

At the present, the experience is a bit different than it was years ago, with all sorts of warnings surrounding cert issues, servers forcing HTTPS, etc. But there is still a problem if there's no HSTS, because the client browser isn't forcing HTTPS requests to the domain in question, i.e. the client browser is still susceptible to sending HTTP requests to the domain. 

That means that if we can get the client to send an HTTP request to the MITM, then the MITM can perform (proxy) an HTTPS request to the server and receive the response (all over TLS), and return an HTTP response (possibly mangled) back to the client. In this case, there's no SSL/TLS being broken. 

So in Ettercap, what I'd like to do with a MITM attack take an HTTP request (packet) from the client, then in an Ettercap filter make a separate HTTPS request to the destination web server, get the HTTPS response, and return it as an HTTP packet to the client. The net effect is that the MITM has intercepted all data in the original request, and returned a response to the client as if it had communicated over HTTPS to the server. 

Is this something feasible in an Ettercap filter (or some other customizable mechanism)?

**Comment by brado** [comment]
Disregard...the use case appears to be out-of-scope for ettercap. Thx for your help.

**Comment by koeppea** [comment]
Mhh maybe not. Your (deleted) explanation makes somewhat sense. It might be possible the same way sslstrip is working: as a plugin. However like sslstrip it would require, that ettercap terminates the TCP connection, which then can only be inspected, but not modified by the filtering engine. Only traffic that is forwarded by ettercap can be modified using filters as of now.

**Comment by brado** [comment]
The use-case is legit -- at its simplest: 

- Forward all network traffic received to every port other than 80 (unencrypted HTTP). 
- Forward all network traffic received to port 80 not matching specific URL patterns (domains, paths, etc.)
- For network traffic received on port 80 and matching specific URL patterns, proxy to the destination web server over HTTPS, receive the response, potentially alter it, and then send the response back to the original client over HTTP. 

Even with a plugin I'm not sure how the proxying would play without being interfered with by the standard ettercap function; i.e. it seems an http request in the plugin would be intercepted by the standard ettercap functionality, which would be undesirable. 

If you think this is feasible, any guidance you can give would be appreciated. If not, knowing that would be appreciated too. Thx. Ettercap rocks -- it would be awesome to accomplish this in ettercap.

**Comment by koeppea** [comment]
I think it's doable. However it reqiuires C network layer programming and an understanding how ettercap works internally. I think I'll put it onto the  TODO list.

**Comment by brado** [comment]
That would be cool...thx.
